packages/daimo-api/src/api/getSwapRoute.ts

```diff
@@ -0,0 +1,139 @@
+import {
+  BigIntStr,
+  EAccount,
+  getSupportedSendCoinByAddress,
+  now,
+  ProposedSwap,
+} from "@daimo/common";
+import { SwapRouter as SwapRouter02 } from "@uniswap/router-sdk";
+import { WETH9 } from "@uniswap/sdk-core";
+import { ADDRESS_ZERO } from "@uniswap/v3-sdk";
+import { Address, Hex, parseAbi } from "viem";
+
+import { UNISWAP_V3_02_ROUTER_ADDRESS } from "../network/uniswapClient";
+import { ViemClient } from "../network/viemClient";
+
+// Direct path length is: 0x (1) + Token (20) + Fee (3) + Token (20) = 44 * 2.
+const SINGLE_POOL_LENGTH_HEX = 88;
+
+// Given an exact amount input of tokenIn, retrieve the best-effort swap path
+// to get tokenOut.
+export async function getSwapQuote({
+  amountInStr,
+  tokenIn,
+  tokenOut,
+  fromAccount,
+  toAddr,
+  chainId,
+  vc,
+}: {
+  amountInStr: BigIntStr;
+  tokenIn: Address;
+  tokenOut: Address;
+  fromAccount: EAccount;
+  toAddr: Address;
+  vc: ViemClient;
+  chainId: number;
+}) {
+  const amountIn: bigint = BigInt(amountInStr);
+  const swapQuote = await vc.publicClient.readContract({
+    address: "0xD49Ad4D3945A655C508498571cFf092D261bf860", // TODO: config
+    abi: parseAbi([
+      "function quote(uint128 amountIn, address tokenIn, address tokenOut) public view returns (uint256 amountOut, bytes memory swapPath)",
+    ]),
+    functionName: "quote",
+    args: [amountIn, tokenIn, tokenOut],
+  });
+
+  const amountOut: bigint = swapQuote[0];
+  const swapPath: Hex = swapQuote[1]; // abi-encoded Uniswap path
+
+  // By default, the router holds the funds until the last swap, then it is
+  // sent to the recipient.
+  // Special case: if outputToken is native, then routerMustCustody is true.
+  // Reference: https://github.com/Uniswap/sdks/blob/main/sdks/universal-router-sdk/src/entities/protocols/uniswap.ts
+  const routerMustCustody = tokenOut === WETH9[chainId].address;
+  const recipient: Address = routerMustCustody ? ADDRESS_ZERO : toAddr;
+
+  const t = now();
+  const cacheUntil = t + 5 * 60; // 5 minutes
+  const execDeadline = t + 10 * 60; // 10 minutes
+
+  const amountOutMinimum = amountOut - amountOut / 100n; // TODO: check amountOutMinimum
+
+  let callData;
+  if (pathIsDirectSwap(swapPath)) {
+    const { tokenIn, fee, tokenOut } = decodeDirectPool(swapPath);
+
+    const exactInputSingleParams = {
+      tokenIn,
+      tokenOut,
+      fee,
+      recipient,
+      amountIn,
+      amountOutMinimum,
+      sqrtPriceLimitX96: 0,
+    };
+    callData = SwapRouter02.INTERFACE.encodeFunctionData("exactInputSingle", [
+      exactInputSingleParams,
+    ]);
+  } else {
+    const exactInputParams = {
+      path: swapPath,
+      recipient,
+      amountIn,
+      amountOutMinimum,
+    };
+    callData = SwapRouter02.INTERFACE.encodeFunctionData("exactInput", [
+      exactInputParams,
+    ]);
+  }
+
+  if (!callData) return null;
+  const execCallData = SwapRouter02.INTERFACE.encodeFunctionData(
```

Comment:
is there any advantage to doing a multi call given we don't take more than one path currently?

Reply:
true the multicall structure is odd -- the uniswap router always returns it and I was just trying to copy the format but ill try without right now and test